Added tcp/hello_protocol acknowledgment commands, verbose logging, and more clear documentation

ajones239 opened the pull request:
I added more commands to the communication between tcp.py and hello_protocol.sh. After the pi receives a command, it responds alerting piman that the command was successfully received. The pi will send another message after the instruction was finished executing. I also changed the tcp documentation to make it less misleading about the tcp.py/hello_protocol.sh exchange.

ajones239 committed:
Added acknowledgment commands, verbose logging, and more clear documentation

markbrogan13 reviewed (approved):
Looks good, good job Aidan

HaozhenWuU reviewed (approved):
Good job.

cjhouser commented on tcp/README.md:
```diff
@@ -1,20 +1,16 @@
 # Why is tcp.py designed like this?
-After reading the 1st header, you might think to yourself: "No handshake? No connection using IP addresses? What kind of TCP protocol is this????"  
-And that is a really good question that won't be answered in this README. However this README can tell that `tcp.py` is designed to like this because its purpose is just to transfer the `/../install/boot/rootfs.tgz` file to the pi.  
+`tcp.py` is designed to like this because its purpose is just to transfer the `/../install/boot/rootfs.tgz` file to the pi.  
```
The path shown here shouldn't start with the root directory.

cjhouser commented on tcp/README.md:
```diff
@@ -1,20 +1,16 @@
-After reading the 1st header, you might think to yourself: "No handshake? No connection using IP addresses? What kind of TCP protocol is this????"  
-And that is a really good question that won't be answered in this README. However this README can tell that `tcp.py` is designed to like this because its purpose is just to transfer the `/../install/boot/rootfs.tgz` file to the pi.  
+`tcp.py` is designed to like this because its purpose is just to transfer the `/../install/boot/rootfs.tgz` file to the pi.  
 But why use `tcp.py` to do this when `/../tftp/tftp.py` already exists to transfer boot files? This is due to `rootfs.tgz` being so large (~500 MB).  
```
Path starts with root directory here, too.

cjhouser commented on tcp/tcp.py:
```diff
@@ -103,8 +112,26 @@ def __process_requests(self, client_socket, client_addr):
-                logger.info("TCP - recieved request {}".format(req))
-                if req == RECV_IS_UNINSTALLED:
+                logger.info("TCP - received request {}".format(req))
+                if req == RECV_RECEIVED_DATE:
```
I like to use [dispatch dictionaries](https://alysivji.github.io/quick-hit-dictionary-dispatch.html) when I need to write big if statements.

ajones239 committed:
Update README.md

Fixed absolute paths

ajones239 committed:
Update tcp.py

Added dictionary dispatch for commands where the only server action is logging. I left the commands that result in server action inside the original if statement.

ajones239 commented:
All absolute paths in the documentation are now relative, and I changed all if statements that resulted in only logging to a dictionary dispatch defined with lambda expressions. I left the other if statements where the command results in action by the server as is because I don't think taking those out into separate functions will make the code more clear.

aldogutierrez reviewed (approved):
&ndash; Reviewed changes!

markbrogan13 reviewed (approved):
Reviewed changes, good work

breed commented on tcp/tcp.py:
```diff
@@ -98,12 +108,23 @@ def __process_requests(self, client_socket, client_addr):
                 req = req.strip()
-                logger.info("TCP - recieved request {}".format(req))
+                logger.info("TCP - received request {}".format(req))
                 if req == RECV_IS_UNINSTALLED:
```
is there a reason this isn't in the dictionary?

breed commented on tcp/tcp.py:
```diff
@@ -98,12 +108,23 @@ def __process_requests(self, client_socket, client_addr):
             while req:
                 req = req.strip()
-                logger.info("TCP - recieved request {}".format(req))
+                logger.info("TCP - received request {}".format(req))
```
doesn't this and line 146 mean that all non request lines get printed 2 times?

ajones239 replied:
Yeah good catch. I removed the duplicate log entry.

breed commented on tcp/tcp.py:
```diff
@@ -98,12 +108,23 @@ def __process_requests(self, client_socket, client_addr):
                 req = req.strip()
-                logger.info("TCP - recieved request {}".format(req))
+                logger.info("TCP - received request {}".format(req))
                 if req == RECV_IS_UNINSTALLED:
```
it would be nice to perhaps include these in the dictionary to be consistent.

ajones239 replied:
I could include these in the dictionary, but I would have to create a separate function for the if statements with multiple statements since lambda expressions only allow a single statement. I left them as is because if they were moved to separate functions the code would be in a different place from the rest of the related logic. Should I move those to functions?

breed replied:
it's totally up to you.

ajones239 replied:
I think it is more clear as is. If they were moved to separate functions, client_sock would have to either be passed as parameter (to only be used in two of the functions/lambdas) or added as an instance variable and I would need to add something to handle the break statement.

breed reviewed (changes requested):
this is really nice. just two small changes/suggestions. let me know what you think

ajones239 committed:
Update tcp.py

breed commented on tcp/tcp.py:
```diff
@@ -119,8 +139,10 @@ def __process_requests(self, client_socket, client_addr):
-                    pass
-                    #print("TCP - not supported request")
+                    try:
+                        req_dict[req]() # log changed state if command is valid and not handled above
```
i thought i had mentioned this before, but i don't see it in the history. can this exception handling be replace with:

```
if req in req_dict:
    req_dict[req]()
```

ajones239 replied:
Yes. It has been updated.

ajones239 committed:
Update tcp.py

Removed try block on line 141, switched to if statement

breed reviewed (approved):
thank you for accomodating all the suggestions!
